### ISSUE TITLE
# of worker processes keeps increasing

### ISSUE THREAD

**Issue by jfeltesse-mdsol** [puma/puma#966]
Hi,

Using puma 3.4.0, ruby 2.3.0, rails 5 master (ref 4769fc4) on Ubuntu 14.04.

The `config/puma.rb` file looks like this:

```ruby
workers ENV.fetch("PUMA_WORKERS") { 2 }
threads_count = ENV.fetch("PUMA_THREADS") { 10 }.to_i
threads threads_count, threads_count * 2

port        ENV.fetch("HTTP_PORT") { 3000 }
environment ENV.fetch("RAILS_ENV") { "development" }

preload_app!

# see https://github.com/puma/puma/commit/448c022
force_shutdown_after 5

on_worker_boot do
 ActiveSupport.on_load(:active_record) do
   ActiveRecord::Base.establish_connection
 end
end

before_fork do
  ActiveRecord::Base.connection_pool.disconnect!
end
```

we launch the app using the following command (the env variables in the config file above are not set so it's using the fallback values (edit: with the exception of the RAILS_ENV one set to "production")):
```
bundle exec rails server -P pids/puma.pid
```

The way I understand this, it should create 2 workers that would spin up 10 threads and be allowed to grow to 20 threads each.
The cluster itself has its own process so we logically, right after deploying, get 3 puma related processes.
The problem is after a while and some access I end up with something like this:

```
$ ps x | grep puma
 51345 ?        Sl     0:07 puma 3.4.0 (tcp://0.0.0.0:3000) [2016-04-21_00-57-21]
 51612 ?        Sl     0:01 puma: cluster worker 0: 51345 [2016-04-21_00-57-21]
 51631 ?        Sl     0:01 puma: cluster worker 1: 51345 [2016-04-21_00-57-21]
 53849 ?        Sl     0:07 puma: cluster worker 0: 51345 [2016-04-21_00-57-21]
 55646 ?        Sl     0:14 puma: cluster worker 1: 51345 [2016-04-21_00-57-21]
 58487 ?        Sl     0:01 puma: cluster worker 0: 51345 [2016-04-21_00-57-21]
 58836 ?        Sl     0:09 puma: cluster worker 0: 51345 [2016-04-21_00-57-21]
 71431 ?        Sl     0:05 puma: cluster worker 1: 51345 [2016-04-21_00-57-21]
 73312 ?        Sl     0:08 puma: cluster worker 0: 51345 [2016-04-21_00-57-21]
 74331 ?        Sl     0:02 puma: cluster worker 1: 51345 [2016-04-21_00-57-21]
 76710 ?        Sl     0:25 puma: cluster worker 1: 51345 [2016-04-21_00-57-21]
 78787 ?        Sl     0:06 puma: cluster worker 0: 51345 [2016-04-21_00-57-21]
 80857 ?        Sl     0:01 puma: cluster worker 0: 51345 [2016-04-21_00-57-21]
 82359 ?        Sl     0:01 puma: cluster worker 0: 51345 [2016-04-21_00-57-21]
 83776 ?        Sl     0:01 puma: cluster worker 0: 51345 [2016-04-21_00-57-21]
 85847 ?        Sl     0:06 puma: cluster worker 0: 51345 [2016-04-21_00-57-21]
 87966 ?        Sl     0:00 puma: cluster worker 1: 51345 [2016-04-21_00-57-21]
 88829 pts/2    S+     0:00 grep --color=auto puma
```

Eventually it eats up all the memory and everything freezes.

The only thing I can imagine could be "uncharted" territory would be we're using ActionCable on one page which is being kept open.

Any pointers?

**Comment by evanphx** [comment]
Well, that's interesting. By "after a while" do you mean the app is just sitting there or are you trying to restart it?

**Comment by jfeltesse-mdsol** [comment]
> By "after a while" do you mean the app is just sitting there or are you trying to restart it?

The app is in testing now so imagine about 5~10 users trying things out for an hour. We're not trying to restart it unless we have to (new code to deploy etc).
Then SSH'ing into the instance shows the system as the log above.
Also, sending TERM to the individual workers does not work. It's KILL or nothing.

We have a super hacky quick n' dirty fix that goes like:
```ruby
on_worker_fork do
  master = ObjectSpace.each_object(Puma::Cluster).first if defined?(Puma::Cluster)
  workers = master&.instance_variable_get(:@workers)
  dead_worker_indexes = workers ? ((0...@options[:workers]).to_a - workers.map(&:index)) : []

  dead_worker_indexes.each do |dead_worker_index|
    list_puma_processes_cmd = "ps -ef | grep -E 'puma: cluster worker [#{dead_worker_index}]' | grep -v grep"
    puts `#{list_puma_processes_cmd}`.split("\n")
    pids = `#{list_puma_processes_cmd} | awk '{print $2;}'`.split("\n")
    puts pids.inspect
    pids.each do |pid|
      puts "sending KILL to puma cluster worker process #{pid}"
      puts `kill -s KILL #{pid}`
    end
  end
end
```

It *seems* that puma thinks a worker died and then tries to spawn a new one to replace it.

Let me know if you need more information or if you want me to try something out.

**Comment by evanphx** [comment]
Is there anything in the logs about trying to kill timeout workers maybe?

**Comment by jfeltesse-mdsol** [comment]
Is there a way to get detailed info in the "puma.log"?
(we set it by doing `stdout_redirect 'log/puma.log', 'log/puma.log', true` in the config file)

Right now with our hack (mostly the same as above, more logging):
```ruby
on_worker_fork do
    Rails.logger.info "A new Puma worker has started"
    master = ObjectSpace.each_object(Puma::Cluster).first if defined?(Puma::Cluster)
    workers = master&.instance_variable_get(:@workers)
    dead_worker_indexes = workers ? ((0...@options[:workers]).to_a - workers.map(&:index)) : []

    dead_worker_indexes.each do |dead_worker_index|
      list_puma_processes_cmd = "ps -ef | grep -E 'puma: cluster worker [#{dead_worker_index}]' | grep -v grep"
      list_puma_processes_cmd_result = `#{list_puma_processes_cmd}`.split("\n")
      Rails.logger.info "Currently running processes info: #{list_puma_processes_cmd_result}"
      pids = `#{list_puma_processes_cmd} | awk '{print $2;}'`.split("\n")
      Rails.logger.info "Pids of currently running processes: #{pids.inspect}"
      pids.each do |pid|
        Rails.logger.warn "sending KILL to puma cluster worker process #{pid}"
        Rails.logger.warn `kill -s KILL #{pid}`
      end
    end
  end
```

I get this kind of stuff:

**puma.log**
```
=== puma startup: 2016-07-26 07:43:06 +0000 ===
[432911] - Worker 0 (pid: 433093) booted, phase: 0
[432911] - Worker 1 (pid: 433121) booted, phase: 0
[432911] - Worker 0 (pid: 434862) booted, phase: 0
[432911] - Worker 1 (pid: 435737) booted, phase: 0
```

**rails' production.log**
```
2016-07-26T07:43:06.952007Z: A new Puma worker has started
2016-07-26T07:43:07.002751Z: Currently running processes info: []
2016-07-26T07:43:07.025638Z: Pids of currently running processes: []
2016-07-26T07:43:07.043337Z: Currently running processes info: []
2016-07-26T07:43:07.065758Z: Pids of currently running processes: []
2016-07-26T07:43:07.067256Z: A new Puma worker has started
2016-07-26T07:43:07.138596Z: Currently running processes info: []
2016-07-26T07:43:07.162656Z: Pids of currently running processes: []

2016-07-26T07:57:22.880862Z: A new Puma worker has started
2016-07-26T07:57:22.932615Z: Currently running processes info: ["polybus  433093 432911  0 07:43 ?        00:00:04 puma: cluster worker 0: 432911 [2016-07-26_07-38-33] "]
2016-07-26T07:57:22.952576Z: Pids of currently running processes: ["433093"]
2016-07-26T07:57:22.952628Z: sending KILL to puma cluster worker process 433093

2016-07-26T08:04:08.034619Z: A new Puma worker has started
2016-07-26T08:04:08.081598Z: Currently running processes info: ["polybus  433121 432911  0 07:43 ?        00:00:03 puma: cluster worker 1: 432911 [2016-07-26_07-38-33] "]
2016-07-26T08:04:08.108456Z: Pids of currently running processes: ["433121"]
2016-07-26T08:04:08.108499Z: sending KILL to puma cluster worker process 433121

2016-07-26T08:11:12.402669Z: A new Puma worker has started
2016-07-26T08:11:12.454752Z: Currently running processes info: ["polybus  434862 432911  0 07:57 ?        00:00:03 puma: cluster worker 0: 432911 [2016-07-26_07-38-33] "]
2016-07-26T08:11:12.481565Z: Pids of currently running processes: ["434862"]
2016-07-26T08:11:12.481612Z: sending KILL to puma cluster worker process 434862

2016-07-26T08:20:56.201496Z: A new Puma worker has started
2016-07-26T08:20:56.242872Z: Currently running processes info: ["polybus  435737 432911  0 08:04 ?        00:00:04 puma: cluster worker 1: 432911 [2016-07-26_07-38-33] "]
2016-07-26T08:20:56.274206Z: Pids of currently running processes: ["435737"]
2016-07-26T08:20:56.274274Z: sending KILL to puma cluster worker process 435737
```

**Comment by evanphx** [comment]
Oh, I'm pretty sure using the same file for both stdout and stderr doesn't work right :( Can you send them to different files?

**Comment by jfeltesse-mdsol** [comment]
sure, let me try this and get back to you in a few days with the findings.

**Comment by jfeltesse-mdsol** [comment]
Nothing interesting appeared in the stdout or stderr logs but meanwhile we discovered that in the code there was a method being called frequently that would execute `ActiveRecord::Base.establish_connection` all the time. 
Changing this bit to `ActiveRecord::Base.connection_pool.with_connection do |c| ... end` made the problem disappear, at least for the past few days.

Closing this for now, will reopen and ping back if the issue continues and we have more detailed info to provide.
Thanks!!
